design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+# Proposal: Define requirements new traits to qualify and enhance the Conan dependencies model
+
+
+| **Status**        |                                                   |
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [026](https://github.com/conan-io/tribe/pull/26)  |
+| **Submitted**     | 2021-07-06                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+Introduce the concept of "requirements traits", to better model the relations between packages. Requirement traits are specifiers of the Conan ``requires`` that can be defined in recipes. These traits can allow defining the current Conan 1.X ``build_requires`` behavior with some ``requires`` traits.
+
+This proposal does not affect ``python_requires``, because of their nature and evaluation at recipe parsing time, they are very different to regular ``requires`` and ``build_requires``, and they would not benefit from this proposal.
+
+Requirement traits will define things like headers inclusions, library linkage, runtime dependency, if the headers visibility should propagate downstream, if the require should propagate downstream (and can conflict) or not, if it is a build tool or a test library, etc.
+
+Package recipes will keep defining their direct dependencies in the same way they were doing in Conan 1.X, but with the possibility of specifying more traits (might not be necessary in many cases, the defaults should be fine). But recipes will have access, once the graph is evaluated, to the ``requires`` of all their dependencies, both direct(declared) and indirect ones (transitive), describing the relation they have with those dependencies with these traits.
+
+
+## Motivation
+
+Conan's main declaration of dependencies is the ``requires`` relationship. Mostly 2 variants of it are also available in Conan 1.X: ``build_requires`` that aim to express a dependency to a build tool, like cmake, and ``private`` which is discouraged for most cases.
+
+But this model does not provide enough information to describe some specifics of the C/C++ compilation model, for example in the general case, when one library is linking another library, the consumers of the former shouldn’t have visibility over the headers of the later one. A ``requires`` trait called ``transitive_headers`` could control this visibility.
+It seems the potential combinations of dependencies usage is too high to consider approaches like the current one, defining high level names as ``build_requires``, ``test_requires``, ``requires_shared``, etc. does not scale.
```

Comment:
agreed, but potentially some separation reduces the cases where the options are in conflict. e.g. does it make sense to have transitive_headers for a build requires?

Reply:
Initially, with the current model it doesn't make sense to have a build_require containing headers, because that build_require lives in the build context, which can be different platform than the host context, and headers will need to compile and run in the host context. The only possible exceptions are compilers/ndks, but as they don't use the typical build-system library inclusion mechanism (there is no need for a ``find_package(gcc)`` or ``find_package(android_ndk)``, not sure yet about the necessity of modeling these headers.

Reply:
I guess I have the desire to filter out invalid settings via the function interface...but I get it's challenging.